engineering/conventions.md

```diff
@@ -79,6 +80,18 @@ Your application will expose one or more binaries with a CLI.
 We prefer [GNU-style command line options](http://www.catb.org/esr/writings/taoup/html/ch10s05.html).
 That is, double dash for long options and single dash for short options (single letter). Include `--help` and `--version`. As a general rule, define always long flags, and then optionally short flags for tools that are often used interactively. When defining flags it is worth to consider matching behavior of widely used flags. For example, if `--all` is present, `-a` should be its shorthand. `--quiet` as a default instead of `--silent`. `--file/-f` for a file argument, `--output/-o` for output path, etc.
 
+## Logging
+
+* Use structured logging.
+* By default, output pretty logs when running in a terminal, output JSON when not.
+* Allow changing logging level and format explicitly through environment variables (`LOG_LEVEL`, `LOG_FORMAT`).
+* Use the following keys where they apply:
+  * `time` for log timestamp.
```

Comment:
Regarding the timestamp: closest to what I found for `strftime`: `%Y-%m-%dT%k:%M:%S.%f000%z`
It still is not exact because the timezone does not include the  double colon after hours.

Reply:
It seems datetime + pytz will do:

```
import pytz
import datetime
d = datetime.datetime.utcnow()
d = d.replace(tzinfo=pytz.UTC)
d.isoformat()
```

outputs: `2018-07-27T15:05:06.649012+00:00`

Reply:
Oh, you cannot put arbitrary code there I guess. Note that a `Z` literal should be valid instead of `+00:00` if time is UTC.

Reply:
Anyway, we should do it in a way that is compatible with Kibana, and I'm still not sure that the current approach is valid. So let's proceed as is and specify datetime format more strictly once we find out.

Reply:
ElasticSearch uses this:
https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html#strict-date-time
http://www.joda.org/joda-time/apidocs/org/joda/time/format/ISODateTimeFormat.html#dateOptionalTimeParser--

It is forgiving in some situations. For example, time is optional and offset might be specified with colon or not:

```
ISODateTimeFormat.dateOptionalTimeParser.parseDateTime("2018-08-02T09:37:13.639950482Z")
res5: DateTime = 2018-08-02T11:37:13.639+02:00

@
ISODateTimeFormat.dateOptionalTimeParser.parseDateTime("2018-08-02T09:37:13.639950482+02:00")
res6: DateTime = 2018-08-02T09:37:13.639+02:00

@
ISODateTimeFormat.dateOptionalTimeParser.parseDateTime("2018-08-02T09:37:13.639950482+0200")
res7: DateTime = 2018-08-02T09:37:13.639+02:00
```

@vmarkovtsev So the format you mention should be ok.

Reply:
Added clarification about format.